### PR TITLE
feat(signals): add signalStore and signalStoreFeature

### LINKED ISSUE #4000: @ngrx/signals: Add initial `signalStore` implementation
### Information

Add initial `signalStore` and `signalStoreFeature` implementation.

### I would be willing to submit a PR to fix this issue

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by markostanimirovic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Closes #4000

## What is the new behavior?

`signalStore` and `signalStoreFeature` functions are added to the `@ngrx/signals` package.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by brandonroberts — APPROVED** [review]

File: modules/signals/src/signal-store-feature.ts [hunk 1]
```diff
@@ -0,0 +1,224 @@
+import {
+  EmptyFeatureResult,
+  MergeFeatureResults,
+  SignalStoreFeature,
+  SignalStoreFeatureResult,
+} from './signal-store-models';
+
+export function signalStoreFeature<F1 extends SignalStoreFeatureResult>(
```
Comment (timdeschryver) [inline modules/signals/src/signal-store-feature.ts:8]: Can these overload types be refactored with variadic tuple types?

Comment (markostanimirovic) [reply modules/signals/src/signal-store-feature.ts:8]: No, it's not possible. Every next feature needs the type of the previous one, similar to [RxJS `pipe`](https://github.com/ReactiveX/rxjs/blob/master/src/internal/util/pipe.ts#L4).

File: modules/signals/spec/signal-store-feature.spec.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    return signalStoreFeature(
+      withCustomFeature1(),
+      withMethods(({ foo, baz }) => ({
+        bar: (value: number) => value,
+        m: () => foo() + baz() + 3,
+      }))
+    );
+  }
+
+  function withCustomFeatureWithInput<_>() {
```
Comment (timdeschryver) [inline modules/signals/spec/signal-store-feature.spec.ts:34]: Does this generic has a meaning?

```suggestion
  function withCustomFeatureWithInput() {
```

File: modules/signals/spec/signal-store-feature.spec.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    return signalStoreFeature(
+      {
+        state: type<{ foo: string }>(),
+        signals: type<{ s: Signal<number> }>(),
```
Comment (timdeschryver) [inline modules/signals/spec/signal-store-feature.spec.ts:38]: I got a question about the usage of `type`.
If I understand this correctly, the `type` is used here to make the store typesafe? In this case, this feature expects that it's used within a store that has a `foo` state property, and a `s` signal property.

**Review by timdeschryver — APPROVED** [review]
I like the different files for each feature 🤩
As mentioned earlier, it seems like my WebStorm fails to infer some types, while VSCode can.

image

File: modules/signals/spec/signal-store-feature.spec.ts (hunk 2 above)
Comment (markostanimirovic) [reply modules/signals/spec/signal-store-feature.spec.ts:34]: I noticed a weird typing bug when custom features with input are used in complex use cases.

[Here is the reproduction.](https://github.com/markostanimirovic/ngrx-signal-store-playground/blob/main/src/tests/test2.ts#L20) If we remove `<_>` from `withY`, the compilation error will be thrown in SignalStores that use `withY` feature although the generic isn't used anywhere. 👀

This is a rare case, but I'll document this behavior and there will be a recommendation in the docs.

File: modules/signals/spec/signal-store-feature.spec.ts (hunk 3 above)
Comment (markostanimirovic) [reply modules/signals/spec/signal-store-feature.spec.ts:38]: Exactly!

**Comment by timdeschryver** [comment]
Congrats on closing #4000! 🚀

**Comment by Maciejlys** [comment]
When can we expect release of this? I assume this will be a npm package ngrx/signals?

**Pull request merged**